pymmcore_plus/core/_mmcore_plus.py

```diff
@@ -43,7 +47,33 @@
 _CHANNEL_REGEX = re.compile("(chan{1,2}(el)?|filt(er)?)s?", re.IGNORECASE)
 
 
+@contextmanager
+def _blockSignal(obj, signal):
+    if isinstance(signal, SignalInstance):
+        signal.block()
+        yield
+        signal.unblock()
+    else:
+        obj.blockSignals(True)
+        yield
+        obj.blockSignals(False)
+
+
+_instance = None
+
+
 class CMMCorePlus(pymmcore.CMMCore):
+    lock = RLock()
+
+    @classmethod
+    def instance(
+        cls, mm_path=None, adapter_paths: ListOrTuple[str] = ()
+    ) -> CMMCorePlus:
+        global _instance
+        if _instance is None:
+            _instance = cls(mm_path, adapter_paths)
+        return _instance

```

Comment:
```suggestion
    __instance = None

    @classmethod
    def instance(
        cls, mm_path=None, adapter_paths: ListOrTuple[str] = ()
    ) -> CMMCorePlus:
        if cls.__instance is None:
            cls.__instance = cls(mm_path, adapter_paths)
        return cls.__instance
```

can we try this as a class attribute rather than a module level global?

Reply:
unfortunately seems like no. See description in commit: https://github.com/tlambert03/pymmcore-plus/pull/73/commits/2bde547f0cf56efc71927d11b05513ec43d853dd

> Has to be this way (single underscore global variable) because
swig seems to prevent us from modifying class attributes, and
double underscore triggers name mangling in a class method:
https://stackoverflow.com/a/59066258/835607

Reply:
concretely:
```python
import pymmcore

class plus(pymmcore.CMMCore):
    __instance = None

    @classmethod
    def instance(cls):
        if cls.__instance is None:
            cls.__instance = cls()
        return cls.__instance
plus.instance()
```
gives

```
Traceback (most recent call last):
  File "/home/ian/Documents/oss/micro/pymmcore-plus/instance.py", line 11, in <module>
    plus.instance()
  File "/home/ian/Documents/oss/micro/pymmcore-plus/instance.py", line 9, in instance
    cls.__instance = cls()
AttributeError: cannot modify read-only attribute 'plus._plus__instance'
```

Reply:
Ohh indeed. Makes sense